### PR TITLE
[RFC] Functorch to automatically compute per-sample gradients

### PR DESCRIPTION
**Pull request by alexandresablayrolles**
**TL;DR** We can use Functorch to compute per sample gradients automatically.

This PR proposes to use Functorch to compute per sample gradients. The "meat" is in grad_sample/functorch.py. It basically does a mini-"per sample" forward/backward on the layer at hand, with the tricks that outputs.backward(backprops) is equivalent to (outputs * backprops).sum().backward().

## Open questions

1. How do we validate the modules? It used to be clear that any module which grad_sampler was not implemented was not OK but now it's different. Should we move from a model where everything not allowed was forbidden, to a model where everything not forbidden is allowed.
2. Do we want to provide clean support for activation/de-activation of GRAD_SAMPLERS or is it ok to just del GRAD_SAMPLERS[module]?

### CODE REVIEW

File: opacus/tests/grad_samples/conv1d_test.py [hunk 1]
```diff
@@ -48,7 +49,8 @@ def test_conv1d(
         dilation: int,
         groups: int,
     ):
-
+        if nn.Conv1d in GradSampleModule.GRAD_SAMPLERS:
```
Comment (ffuuugor) [inline opacus/tests/grad_samples/conv1d_test.py:52]: huh?

File: opacus/grad_sample/functorch.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+        return loss
+
+    ft_compute_grad = grad(compute_loss_stateless_model)
+
+    layer.ft_compute_sample_grad = vmap(ft_compute_grad, in_dims=(None, 0, 0))
+    layer.func_params = params
+
+
+def compute_per_sample_gradient(layer, activations, backprops):
```
Comment (ffuuugor) [inline opacus/grad_sample/functorch.py:21]: nit: I'd call it smth like `ft_compute_per_sample_gradient` or otherwise that would indicate that it's functorch-based functionality.

And probably explicitly check that `prepare_layer` has been called on the input `layer`

File: opacus/grad_sample/grad_sample_module.py [hunk 3]
```diff
@@ -219,20 +220,22 @@ def add_hooks(
             self.autograd_grad_sample_hooks = self._module.autograd_grad_sample_hooks
 
         for _module_name, module in trainable_modules(self._module):
-            if type(module) in self.GRAD_SAMPLERS:
-                self.autograd_grad_sample_hooks.append(
-                    module.register_forward_hook(self.capture_activations_hook)
-                )
+            if not type(module) in self.GRAD_SAMPLERS: 
+                prepare_layer(module)
```
Comment (ffuuugor) [inline opacus/grad_sample/grad_sample_module.py:224]: question: can we call `prepare_layer` from within `compute_per_sample_gradient`? To avoid extra preparation step

Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py:224]: Oh, and we probably want to add extra bits to '_close()' method to remove extra attributes we add in `prepare_layer`

**Review by ffuuugor — COMMENTED** [review]
That's amazing! a) the fact that it works and b) the fact that it looks so neat and elegant is just great.

One question I have about making this a part of classic `GradSampleModule` - are there any edge cases where functorch could do things wrong? Like RNNs, tied weights, GANs, etc (that's just a list of things I remember causing issues in the past, I don't necessarily see a reason they shouldn't work with functorch)

> How do we validate the modules?

I think what you did there is good as a first step. We don't have to solve all of our problems at once, and the problem of custom layers doing strange transforms existed before. 

I think in the future we can provide a utility function that mimics our grad_sample tests to give clients an opportunity to check their models, but for now I think we shouldn't try to kill two birds with one stone.

> Do we want to provide clean support for activation/de-activation of GRAD_SAMPLERS or is it ok to just del GRAD_SAMPLERS[module]?

What could be the reason for a client to disable custom grad_sampler?

File: opacus/grad_sample/grad_sample_module.py (hunk 3 above)
Comment (alexandresablayrolles) [reply opacus/grad_sample/grad_sample_module.py:224]: > question: can we call `prepare_layer` from within `compute_per_sample_gradient`? To avoid extra preparation step

Sounds a bit wasteful no? Especially since GSM already does a preparation step to add hooks.

**Review by ffuuugor — COMMENTED** [review]
Looks great!
Changes on `GradSampleModule` level LGTM, now let's make them available in `PrivacyEngine` API. I think it's best to reuse 'grad_sample_mode' parameter for that, adding the 3rd possible value there.

Regarding the tests,
1) For the "module not supported" tests in `tests/grad_sample_module_test.py` I would just switch False to True - we want to make sure GSM doesn't raise a validation exception for the unknown modules now
2) I think it's a good idea to update `tests/gradient_accumulation_test.py` as well. I'm not sure what behaviour functorch provides, but we want to know it and document it.
3) `tests/module_validator_test.py` should stay as it is, we still don't want BatchNorm
4) It'd be good to add another use case to `tests/privacy_engine_test.py` covering custom modules and non-supported core modules (maybe something like `nn.Transformer*`)

Again, amazing diff, really a step forward for opacus

File: opacus/grad_sample/grad_sample_module.py [hunk 4]
```diff
@@ -124,7 +129,12 @@ def __init__(
         self.hooks_enabled = False
         self.batch_first = batch_first
         self.loss_reduction = loss_reduction
-        self.add_hooks(loss_reduction=loss_reduction, batch_first=batch_first)
+        self.add_hooks(
+            loss_reduction=loss_reduction,
+            batch_first=batch_first,
+            force_functorch=force_functorch,
+        )
+        self.force_functorch = force_functorch
```
Comment (karthikprasad) [inline opacus/grad_sample/grad_sample_module.py:137]: teeny tiny nit: move this line below L131

File: opacus/grad_sample/grad_sample_module.py [hunk 5]
```diff
@@ -257,6 +275,11 @@ def remove_hooks(self) -> None:
             delattr(self, "autograd_grad_sample_hooks")
             delattr(self._module, "autograd_grad_sample_hooks")
 
+        # Remove functorch hooks
+        for _module_name, module in trainable_modules(self._module):
+            if not type(module) in self.GRAD_SAMPLERS:
+                delattr(module, "ft_compute_sample_grad")
```
Comment (karthikprasad) [inline opacus/grad_sample/grad_sample_module.py:281]: Shouldn't this be done for all layers when force_functorch is True?

File: opacus/grad_sample/functorch.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+from functorch import make_functional, vmap, grad
+from opacus.layers.dp_rnn import RNNLinear
+
+
+def prepare_layer(layer, batch_first=True):
```
Comment (karthikprasad) [inline opacus/grad_sample/functorch.py:5]: Please add documentation explaining what this function does.

File: opacus/grad_sample/grad_sample_module.py [hunk 7]
```diff
@@ -444,9 +471,10 @@ def is_supported(cls, module: nn.Module) -> bool:
         Returns:
             ``True`` if grad sampler is found, ``False`` otherwise
         """
-        return type(module) in cls.GRAD_SAMPLERS or isinstance(
-            module, (DPRNNBase, DPRNNCellBase)
-        )
+        return True
+        # return type(module) in cls.GRAD_SAMPLERS or isinstance(
+        #     module, (DPRNNBase, DPRNNCellBase)
+        # )
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:477]: 🦕

File: opacus/layers/dp_rnn.py [hunk 8]
```diff
@@ -630,6 +630,24 @@ def initialize_cells(self):
         self.set_rename_map(rename_map)
         return cells
 
+    def parameters(self, recurse=True):
+        """
+        Redefining parameters() as the RenameParamsMixin confuses it
```
Comment (karthikprasad) [inline opacus/layers/dp_rnn.py:635]: sorry, confuses what?

File: opacus/tests/grad_samples/conv1d_test.py [hunk 9]
```diff
@@ -21,6 +21,7 @@
 from hypothesis import given, settings
 
 from .common import GradSampleHooks_test, expander, shrinker
+from opacus.grad_sample import GradSampleModule
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/tests/grad_samples/conv1d_test.py:24]: 🦖

File: requirements.txt [hunk 10]
```diff
@@ -1,3 +1,4 @@
+functorch
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply requirements.txt:1]: The main thing that held us back from bumping up the torch dependency was csprng. Sorry, I'm not familiar with the latest take on that, are we okay with opacus sans csprng?

File: opacus/grad_sample/functorch.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        output = flayer(params, batched_activations)
+        loss = (output * batched_backprops).sum()
+
+        return loss
+
+    ft_compute_grad = grad(compute_loss_stateless_model)
+    layer.ft_compute_sample_grad = vmap(ft_compute_grad, in_dims=(None, 0, 0))
+
+
+def ft_compute_per_sample_gradient(layer, activations, backprops):
```
Comment (karthikprasad) [inline opacus/grad_sample/functorch.py:27]: ditto

File: opacus/grad_sample/functorch.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+from functorch import make_functional, vmap, grad
```
Comment (karthikprasad) [inline opacus/grad_sample/functorch.py:1]: Wow! So elegant! <3

File: opacus/grad_sample/grad_sample_module.py [hunk 13]
```diff
@@ -211,6 +225,7 @@ def add_hooks(
                 ``[K, batch_size, ...]``
             loss_reduction: Indicates if the loss reduction (for aggregating the gradients)
                 is a sum or a mean operation. Can take values "sum" or "mean"
+            force_functorch: If set to ``True``, will use functorch to compute per sample gradients.
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:228]: Also worth explaining, either here or on the website/readme, as to why one should choose one over the other.

File: opacus/grad_sample/grad_sample_module.py [hunk 14]
```diff
@@ -352,7 +375,11 @@ def capture_backprops_hook(
             loss_reduction=loss_reduction,
             batch_first=batch_first,
         )
-        grad_sampler_fn = self.GRAD_SAMPLERS[type(module)]
+        if not self.force_functorch and type(module) in self.GRAD_SAMPLERS:
+            grad_sampler_fn = self.GRAD_SAMPLERS[type(module)]
+        else:
+            grad_sampler_fn = ft_compute_per_sample_gradient
+
```
[the remark this answers is not in the document]
Comment (alexandresablayrolles) [reply opacus/grad_sample/grad_sample_module.py:382]: It's all in `rearrange_grad_samples` that puts the batch first if needed and scales backprops appropriately for `loss_reduction`

File: requirements.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (alexandresablayrolles) [reply requirements.txt:1]: > are we okay with opacus sans csprng?

csprng has been an optional dependency since fc23b23ed19a9d0bdc1d1a7a55bcef3a5904b36f (which means you could have functorch+pytorch1.11 without csprng even though it's not ideal). That being said, I think @romovpa fixed the csprng dependency on older version of Pytorch but not 100% sure.

> given that functorch requires torch>=1.11, do we want to update our dependency too?

If we put functorch, wouldn't pip automatically apply this constraint to the torch version?

File: opacus/layers/dp_rnn.py (hunk 8 above)
Comment (alexandresablayrolles) [reply opacus/layers/dp_rnn.py:635]: For some reason, RenameParamsMixin gives parameters() for both the DPRNN and the RNNLinear, which are the same. This is a way around it but there might be something cleaner?

File: opacus/grad_sample/grad_sample_module.py (hunk 5 above)
Comment (alexandresablayrolles) [reply opacus/grad_sample/grad_sample_module.py:283]: Done. I'm thinking, we can also just check dynamically if the attribute is there and remove it?

Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:287]: Thanks. That works too.

File: requirements.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (karthikprasad) [reply requirements.txt:1]: > csprng has been an optional dependency since fc23b23

That's right! IIUC, csprng is optional dependency, but it can still be used if needed. csprng is only supported until torch 1.8, and the current opacus is compatible with torch 1.8.
However, if we change the torch version from `torch>=1.8` to `torch>=1.11` in L3 in this file, opacus will no longer be useable with torch 1.8 and consequently with csprng.
I agree that @romovpa would have the best context here.

File: opacus/layers/dp_rnn.py (hunk 8 above)
Comment (karthikprasad) [reply opacus/layers/dp_rnn.py]: I don't know if there is a cleaner way; this implementation is clean enough TBH. My question was mostly around what "it" refers to in the docstring; elaborating it would be useful :)

Comment (alexandresablayrolles) [reply opacus/layers/dp_rnn.py]: I found a way around that (basically not triggering grad samples on DPRNN/DPLSTM as it is done on their LSTMLinear), not needed anymore!

**Review by ffuuugor — APPROVED** [review]
Love this!
Approving, as all my comments are small nits.

One more thing we want to have is additional test in "tests/privacy_engine_test.py" to confirm that functorch produces correct results.
It doesn't have to be a part of this PR - can you pls create a new issue so that we follow up on this later?

File: opacus/grad_sample/grad_sample_module.py [hunk 15]
```diff
@@ -374,10 +409,13 @@ def is_supported(cls, module: nn.Module) -> bool:
         Returns:
             ``True`` if grad sampler is found, ``False`` otherwise
         """
-        return type(module) in cls.GRAD_SAMPLERS or isinstance(
-            module, (DPRNNBase, DPRNNCellBase)
+        warnings.warn(
+            "GradSampleModule.is_supported is deprecated, as all layers can now be used with functorch.",
```
Comment (karthikprasad) [inline opacus/grad_sample/grad_sample_module.py:413]: The biggest impacc of this change! Thank you

File: requirements.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (karthikprasad) [reply requirements.txt:1]: Just reopened this thread. Did we reach a conclusion on which torch version to support?
Given that functorch is listed as a requirement, what would happen with when someone tries to use functorch on torch 1.8.1?

**Review by karthikprasad — APPROVED** [review]
